From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phoenix616 <mail@moep.tv>
Date: Sun, 29 Dec 2019 00:53:17 +0100
Subject: [PATCH] Origami: Add options for modifying pigmen farm behaviour

Original code by Phoenix616, licensed under GNU Lesser General Public License v3.0
you can find the original code on https://github.com/Minebench/Origami/

Now updated to work better with Purpur.

Updated to 1.17 by: notsatvrn <satvrn@gmx.com>

diff --git a/src/main/java/io/graphenemc/graphene/GrapheneWorldConfig.java b/src/main/java/io/graphenemc/graphene/GrapheneWorldConfig.java
index 7d8a66e962a6c607482148258a5e4e4bd5560228..f97f42ac0197be21b10d6e6288b31dc88211eacf 100644
--- a/src/main/java/io/graphenemc/graphene/GrapheneWorldConfig.java
+++ b/src/main/java/io/graphenemc/graphene/GrapheneWorldConfig.java
@@ -76,4 +76,19 @@ public class GrapheneWorldConfig {
         final Map<String, Object> value = GrapheneConfig.getMap("world-settings." + worldName + "." + path, null);
         return value.isEmpty() ? fallback : value;
     }
+
+    public double pigmenPortalBoost = 0.0;
+    public boolean pigmenFarmsWaterAi = false;
+    public int pigmenFarmsRadius = 16;
+    public int pigmenFarmsCount = 5;
+    private void pigmenImprovements() {
+        pigmenPortalBoost = getDouble("pigmen.portal-boost", pigmenPortalBoost);
+        pigmenFarmsWaterAi = getBoolean("pigmen.farms.water-ai", pigmenFarmsWaterAi);
+        pigmenFarmsRadius = getInt("pigmen.farms.radius", pigmenFarmsRadius);
+        pigmenFarmsCount = getInt("pigmen.farms.count", pigmenFarmsCount);
+        if (pigmenFarmsWaterAi) {
+            log("Pigmen water AI in farms is enabled");
+            log("Farm detection radius: " + pigmenFarmsRadius + ", count: " + pigmenFarmsCount);
+        }
+    }
 }
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 2ffef070ec016fcfd2b7784f78a64f532772cabc..967c86047e436e729cb8f4099e1e22c98c712e3f 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -335,6 +335,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
     public boolean isTemporarilyActive = false; // Paper
     public boolean spawnedViaMobSpawner; // Paper - Yes this name is similar to above, upstream took the better one
     public boolean fromNetherPortal; // Paper
+    public boolean fromMobFarm = false; // Origami - Detect Mobs from mob farms
     protected int numCollisions = 0; // Paper
     public void inactiveTick() { }
     // Spigot end
@@ -2187,6 +2188,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
                 nbt.putBoolean("Paper.FromNetherPortal", true);
             }
             // Paper end
+            // Origami start
+            if (fromMobFarm) {
+                nbt.putBoolean("Origami.FromMobFarm", true);
+            }
+            // Origami end
             return nbt;
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Saving entity NBT");
@@ -2351,6 +2357,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
                 spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT;
             }
             // Paper end
+            fromMobFarm = nbt.getBoolean("Origami.FromMobFarm"); // Origami - Restore entity's from mob farm status
 
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Loading entity NBT");
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 1b3739b7288cbb95c0c844b9e6261fb123d56d6a..6d65eebbba098f9ead0ec7e98b3fc9b8e48c1189 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -860,7 +860,7 @@ public abstract class Mob extends LivingEntity {
     @Override
     protected final void serverAiStep() {
         ++this.noActionTime;
-        if (!this.aware) { // Paper start - Allow nerfed mobs to jump, float and take water damage
+        if (!this.aware || this.fromMobFarm && this.isInWater()) { // Paper start - Allow nerfed mobs to jump, float and take water damage // Origami - Mobfarm water ai
             if (goalFloat != null) {
                 if (goalFloat.canUse()) goalFloat.tick();
                 this.getJumpControl().tick();
diff --git a/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java b/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java
index c42349d0f6b0025525278295b36f4030f122800e..08ea7d26dafb0b3c1a7038fd156fa0f09e9fff03 100644
--- a/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -68,6 +68,13 @@ public class NetherPortalBlock extends Block {
                     entity.setPortalCooldown();
                     entity.fromNetherPortal = true; // Paper
                     if (world.paperConfig.nerfNetherPortalPigmen) ((net.minecraft.world.entity.Mob) entity).aware = false; // Paper
+                    // Origami start - Pigmen farm improvements
+                    if (world.grapheneConfig.pigmenFarmsWaterAi && world.grapheneConfig.pigmenFarmsCount > 0 && world.grapheneConfig.pigmenFarmsRadius > 0) {
+                        if (world.getEntitiesOfClass(net.minecraft.world.entity.monster.ZombifiedPiglin.class, new net.minecraft.world.phys.AABB(pos).inflate(world.grapheneConfig.pigmenFarmsRadius), e -> true).size() > world.grapheneConfig.pigmenFarmsCount) {
+                            entity.fromMobFarm = true;
+                        }
+                    }
+                    // Origami end
                 }
             }
         }
@@ -95,6 +102,19 @@ public class NetherPortalBlock extends Block {
                 return;
             }
             // Purpur end
+            // Origami start - Pig Zombie portal boost
+            if (entity instanceof net.minecraft.world.entity.monster.ZombifiedPiglin && world.grapheneConfig.pigmenPortalBoost > 0) {
+                net.minecraft.world.phys.Vec3 velocity = new net.minecraft.world.phys.Vec3(entity.getLookAngle().x, 0, entity.getLookAngle().z);
+                if (entity.getDirection().getAxis() != state.getValue(NetherPortalBlock.AXIS)) {
+                    velocity = velocity.yRot(90f); // Rotate around y axis
+                }
+                velocity = velocity
+                        .normalize()
+                        .scale(world.grapheneConfig.pigmenPortalBoost);
+                entity.move(net.minecraft.world.entity.MoverType.SELF, velocity);
+                return;
+            }
+            // Origami end
             // CraftBukkit start - Entity in portal
             EntityPortalEnterEvent event = new EntityPortalEnterEvent(entity.getBukkitEntity(), new org.bukkit.Location(world.getWorld(), pos.getX(), pos.getY(), pos.getZ()));
             world.getCraftServer().getPluginManager().callEvent(event);
